kernel/dw1000.c

```diff
@@ -1867,11 +1906,53 @@ static void dw1000_rx_irq(struct dw1000 *dw)
 	 * Instead, we simply discard the received frame and toggle
 	 * HRBPT as for a normally received frame.
 	 */
-	if (unlikely(rx->rxovrr & DW1000_SYS_STATUS2_RXOVRR)) {
-		dev_err_ratelimited(dw->dev, "RX overrun; aborting receive\n");
-		len = 0;
+
+	status = le32_to_cpu(rx->status);
+
+	/* RX out-of-sync check */
+	if (unlikely(!DW1000_HSRPB_SYNC(status))) {
+		dev_warn_ratelimited(dw->dev, "RX sync lost after %d frames; "
+				    "recovering\n", dw->rx_sync_cnt);
+		goto err_recover2;
+	}
+
+	/* RX overruns check */
+	overruns = le16_to_cpu(rx->evc_ovr);
+	if (unlikely(overruns != dw->rx_overruns)) {
+		dev_warn_ratelimited(dw->dev, "RX overruns (%d) detected; "
+				     "recovering\n",
+				     overruns - dw->rx_overruns);
```

Comment:
I'm pretty sure this subtraction will give the wrong answer when the 12-bit counter wraps around.